core/app.cpp

```diff
@@ -831,66 +836,74 @@ std::string restructured_text_usage() {
+  // let's check if arg starts with a dash
+  auto no_dash_arg = without_leading_dash(arg);
+  if (arg.size() == no_dash_arg.size() || no_dash_arg.empty() || isdigit(no_dash_arg.front()) != 0 ||
+      no_dash_arg.front() == '.') {
+    return nullptr;
+  }
 
-    for (size_t i = 0; i < OPTIONS.size(); ++i)
-      get_matches(candidates, OPTIONS[i], root);
-    get_matches(candidates, __standard_options, root);
+  std::vector<const Option *> candidates;
+  std::string root(no_dash_arg);
 
-    // no matches
-    if (candidates.empty())
-      throw Exception(std::string("unknown option \"-") + root + "\"");
+  for (size_t i = 0; i < OPTIONS.size(); ++i)
+    get_matches(candidates, OPTIONS[i], root);
+  get_matches(candidates, __standard_options, root);
 
-    // return match if unique:
-    if (candidates.size() == 1)
-      return candidates[0];
+  // no matches
+  if (candidates.empty())
+    throw Exception(std::string("unknown option \"-") + root + "\"");
 
-    // return match if fully specified:
-    for (size_t i = 0; i < candidates.size(); ++i)
-      if (root == candidates[i]->id)
-        return candidates[i];
+  // return match if unique:
+  if (candidates.size() == 1)
+    return candidates[0];
 
-    // check if there is only one *unique* candidate
-    const auto cid = candidates[0]->id;
-    if (std::all_of(++candidates.begin(), candidates.end(), [&cid](const Option *cand) { return cand->id == cid; }))
-      return candidates[0];
+  // return match if fully specified:
+  for (size_t i = 0; i < candidates.size(); ++i)
+    if (root == candidates[i]->id)
+      return candidates[i];
 
-    // report something useful:
-    root = "several matches possible for option \"-" + root + "\": \"-" + candidates[0]->id;
+  // check if there is only one *unique* candidate
+  const auto cid = candidates[0]->id;
+  if (std::all_of(++candidates.begin(), candidates.end(), [&cid](const Option *cand) { return cand->id == cid; }))
+    return candidates[0];
 
-    for (size_t i = 1; i < candidates.size(); ++i)
-      root += std::string("\", \"-") + candidates[i]->id + "\"";
+  // report something useful:
+  root = "several matches possible for option \"-" + root + "\": \"-" + candidates[0]->id;
 
-    throw Exception(root);
-  }
+  for (size_t i = 1; i < candidates.size(); ++i)
+    root += std::string("\", \"-") + candidates[i]->id + "\"";
 
-  return nullptr;
+  throw Exception(root);
 }
 
-void sort_arguments(int argc, const char *const *argv) {
-  for (int n = 1; n < argc; ++n) {
-    if (argv[n]) {
-      const Option *opt = match_option(argv[n]);
-      if (opt) {
-        if (n + int(opt->size()) >= argc)
-          throw Exception(std::string("not enough parameters to option \"-") + opt->id + "\"");
-
-        option.push_back(ParsedOption(opt, argv + n + 1));
-        n += opt->size();
-      } else
-        argument.push_back(ParsedArgument(nullptr, nullptr, argv[n]));
+void sort_arguments(const std::vector<std::string> &arguments) {
```

Comment:
What's your level of confidence in the reimplementation of `match_option()` / `sort_arguments()`? It's hard to review from code alone. Could possibly make use of `testing_cpp_cli`: if the testing tool writes out the arguments in sorted order / what options were specified and what their corresponding arguments are, then can run tests with various permutations of argument / option order and make sure that what was parsed was as intended. Could even add tests with non-ASCII "dash" characters.

Reply:
I have tested this manually and using "standard" unit tests with Google Test (as an aside, it's very easy to use Google Test with our codebase since our CMake switch).
It's good idea to tests to the codebase though. I will try to add some tests with `testing_cpp_cli` to ensure correctness.